pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 			Name:      "distributor_forward_samples_total",
 			Help:      "The total number of samples the Distributor forwarded.",
 		}),
+		exemplarsTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace: "cortex",
+			Name:      "distributor_forward_exemplars_total",
+			Help:      "The total number of exemplars the Distributor forwarded.",
+		}),
 		requestLatencyHistogram: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
 			Namespace: "cortex",
 			Name:      "distributor_forward_requests_latency_seconds",
 			Help:      "The client-side latency of requests to forward metrics made by the Distributor.",
 			Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 20, 30},
 		}),
 	}
+
+	f.workerWg.Add(f.cfg.RequestConcurrency)
+	for i := 0; i < f.cfg.RequestConcurrency; i++ {
+		go f.worker()
+	}
+
+	return f
 }
 
-func (r *forwarder) NewRequest(ctx context.Context, tenant string, rules validation.ForwardingRules) Request {
-	return &request{
-		ctx:    ctx,
-		client: &r.client, // http client should be re-used so open connections get re-used.
-		pools:  &r.pools,
+func (f *forwarder) Stop() {
+	close(f.reqCh)
+	f.workerWg.Wait()
+}
 
-		tsByEndpoint: make(map[string]*[]mimirpb.PreallocTimeseries),
+func (f *forwarder) Forward(ctx context.Context, rules validation.ForwardingRules, in []mimirpb.PreallocTimeseries) (TimeseriesCounts, []mimirpb.PreallocTimeseries, chan error) {
+	if !f.cfg.Enabled {
+		errCh := make(chan error)
+		close(errCh)
+		return TimeseriesCounts{}, in, errCh
+	}
 
-		rules:           rules,
-		timeout:         r.cfg.RequestTimeout,
-		propagateErrors: r.cfg.PropagateErrors,
+	notIngestedCounts, toIngest, tsByTargets := f.splitByTargets(in, rules)
+	defer f.pools.putTsByTargets(tsByTargets)
 
-		requests: r.requestsTotal,
-		errors:   r.errorsTotal,
-		samples:  r.samplesTotal,
-		latency:  r.requestLatencyHistogram,
+	var requestWg sync.WaitGroup
+	requestWg.Add(len(tsByTargets))
+	errCh := make(chan error, len(tsByTargets))
+	for endpoint, ts := range tsByTargets {
+		f.newRequest(ctx, endpoint, ts, &requestWg, errCh)
 	}
+
+	go func() {
```

Comment:
I must be missing something. I don't understand why this isn't just code outside of a goroutine.

Reply:
without the goroutine the call to `.Forward()` would block until the forwarding requests have completed because of the `requestWg.Wait()`. Running this in a go routine `.Forward()` returns early, allowing the distributor to submit the write requests to the ingesters concurrently with the forwarding requests being executed, once the write requests to the ingesters have completed the distributor waits for the `errCh` to close in order to block until the forwarding requests are also done and handle potential errors. 

I'll add a comment explaining that.

Reply:
Explained here: 0f1f750a4